### ISSUE TITLE
Memory increase and never released , how to releas the memory ?

### ISSUE THREAD

**Issue by china-qin** [ntop/nDPI#2171]
I write a sample pcap by dpdk , and inspection the packet by nDpi with the ndpiReader sampel changed.

I get * iph or *iph6 , just like the ndpiReader , then call the ndpi_detection_process_packet function to get the L7 app protocol , like this:

   ```
    struct ndpi_flow_input_info *input_info = (struct ndpi_flow_input_info *) malloc(sizeof(struct ndpi_flow_input_info));;
    memset(input_info, '\0', sizeof(input_info)); /* To be sure to set to "unknown" any fields */
    input_info->in_pkt_dir = NDPI_IN_PKT_DIR_UNKNOWN;
    input_info->seen_flow_beginning = NDPI_FLOW_BEGINNING_UNKNOWN;

    struct ndpi_flow_struct *ndpi_flow = (struct ndpi_flow_struct *) malloc(sizeof(struct ndpi_flow_struct));;;
    memset(ndpi_flow, '\0', SIZEOF_FLOW_STRUCT);

    ndpi_detection_process_packet(this->ndpi_struct, ndpi_flow,
                                                      iph ? (uint8_t *) iph : (uint8_t *) iph6,
                                                      packetLength, pPacket->getPacketTimeStamp().tv_sec, nullptr);
    free(input_info);
    free(ndpi_flow);
```

The problem is the memory continue increse and never released. (increase slowly)

When i call the ndpi_detection_process_packet function , need i call some function follow ?

The nDpi version i used is [4.8 Stable] .

**Comment by china-qin** [comment]
```
free(input_info);
free(ndpi_flow);
```
change to 

```
ndpi_free(input_info);
ndpi_flow_free(ndpi_flow);
```
Now , every thing is ok .
And , about over is all correct ?

**Comment by IvanNardi** [comment]
There are lots of issue with that code...
1) never mix `malloc/free` with `ndpi_malloc/ndpi_free`: memory allocated with `malloc` MUST be freed with `free`
2) you never use `input_info` so no need to allocate it. Note that it is fine to not use `input_info`: it is some kind of "advance" feature
3) You must keep the same `ndpi_flow` for all the packets of the same flow. Please note that flow management is up to the application, i.e. up to you

Bottom line: take a closer look at the code in `ndpiReader` or `ndpiSimpleIntegration`; the latter is definitely easier to follow/understand

**Comment by china-qin** [comment]
Realy learned a lot , and thanks a lot . @IvanNardi
